### PR TITLE
Make CMake runs idempotent

### PR DESCRIPTION
**Pull request by maxaehle**
When I run `cmake` and `make` in the following order,

  1. `cmake`
  2. `make`
  3. `cmake`
  4. `make`,

on my system (Ubuntu with GCC 9.4.0, CMake 3.16.3) and with the CMake flags listed below, the second `make` run (step 4) recompiles a lot of source files using a different set of compiler flags than in the first `make` run (step 2). I think this is undesirable because users expect basically the same program to be build no matter how often they have run `cmake` before.

Specifically, I used the following CMake command slightly adapted from the [Dockerfile of the GATE toolkit](https://github.com/OpenGATE/Gate/blob/8dcb0ddc344561053997d4cdd126c3d4f1acfc1d/source/docker/DockerFileGeant#L78-L84):
```bash
cmake path/to/ITK -DCMAKE_CXX_FLAGS:STRING=-std=c++11 -DBUILD_DOCUMENTATION=OFF -DBUILD_EXAMPLES=OFF -DBUILD_TESTING=OFF -DModule_RTK=ON -DRTK_USE_CUDA=OFF -DCMAKE_BUILD_TYPE=Debug -DRTK_BUILD_APPLICATIONS=OFF
```

The changes proposed in this PR achieve the expected "idempotence" of CMake calls for the above set of CMake flags. After these changes, only a single file [itkBuildInformation.cxx.in](https://github.com/InsightSoftwareConsortium/ITK/blob/490b5c6069a790c2bbc50ece3a578fe1c5cdd50b/Modules/Core/Common/src/itkBuildInformation.cxx.in) is recompiled in step 4 (and some targets are relinked). If you consider these changes helpful, we can explore more sets of CMake flags.

Commit 22f622857040ee28d23e38ea95435ceee05e8fbb in this PR targets the third-party dependency HDF5. I am unsure whether my procedure is in accordance with ITK's [guide on updating third-party projects](https://github.com/InsightSoftwareConsortium/ITK/blob/490b5c6069a790c2bbc50ece3a578fe1c5cdd50b/Documentation/Maintenance/UpdatingThirdParty.md). The suggested patch 
[HDF5-COMP-Perform-TEST_LFS_WORKS-in-every-CMake-run.patch.txt](https://github.com/InsightSoftwareConsortium/ITK/files/10352572/HDF5-COMP-Perform-TEST_LFS_WORKS-in-every-CMake-run.patch.txt)
partially reverts HDFGroup/hdf5@4659e43a6b0914e3da6048328cece6e01be95e44.

Please be warned that I am not an expert in CMake or ITK. Also, I did not run any tests on my own.

## PR Checklist
- [X] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [X] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)
- [X] ~~Added test~~ (or <ins>behavior not changed</ins>)
- [X] ~~Updated API documentation~~ (or <ins>API not changed</ins>)
- [X] ~~Added [license](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Utilities/KWStyle/ITKHeader.h) to new files (if any)~~ (no new files)
- [X] ~~Added Python wrapping to new files (if any) as described in [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) Section 9.5~~ (no new files)
- [X] ~~Added [ITK examples](https://github.com/InsightSoftwareConsortium/ITKSphinxExamples) for all new major features (if any)~~ (no new features)

Refer to the [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) for
further development details if necessary.

### CODE REVIEW

File: Modules/ThirdParty/HDF5/src/itkhdf5/config/cmake_ext_mod/ConfigureChecks.cmake [hunk 1]
```diff
@@ -275,7 +275,7 @@ if (MINGW OR NOT WINDOWS)
   else ()
   set(HDF_ENABLE_LARGE_FILE ON)
   endif ()
-  if (HDF_ENABLE_LARGE_FILE AND NOT DEFINED TEST_LFS_WORKS_RUN)
+  if (HDF_ENABLE_LARGE_FILE)
```
Comment (dzenanz) [inline Modules/ThirdParty/HDF5/src/itkhdf5/config/cmake_ext_mod/ConfigureChecks.cmake:278]: Submitting this patch to [HDF5](https://github.com/HDFGroup/hdf5) is desirable.

**Review by dzenanz — COMMENTED** [review]
I like the purpose of this PR. Removing SS2 flag might be an issue.

**Comment by issakomi** [comment]
I have tried to reproduce:
```
cmake -DBUILD_TESTING:BOOL=OFF ../ITK
...
make
...
cmake -DBUILD_TESTING:BOOL=OFF ../ITK
...
make
...
```
There are in fact some strange thing, e.g. after the 2-nd run `-msse2` flag appears, no idea why, it is **not** required on x86_64. 

```
r@deb2:~/Desktop$ diff CMakeFiles-1/ITKConfig.cmake CMakeFiles-2/ITKConfig.cmake 
27c27
< set(ITK_REQUIRED_CXX_FLAGS " -fno-sized-deallocation")
---
> set(ITK_REQUIRED_CXX_FLAGS " -fno-sized-deallocation -msse2")
66c66
< set(ITK_BUILD_DOCUMENTATION "")
---
> set(ITK_BUILD_DOCUMENTATION "OFF")
```

Related to _target hdf5-static_

```
diff CMakeFiles-1/Export/2e2433a34034ef32dec1447bd4e6a2fe/ITKTargets.cmake CMakeFiles-2/Export/2e2433a34034ef32dec1447bd4e6a2fe/ITKTargets.cmake 
436c436
<   INTERFACE_COMPILE_DEFINITIONS "_POSIX_C_SOURCE=200809L;_GNU_SOURCE;_FILE_OFFSET_BITS=64;_LARGEFILE64_SOURCE;_LARGEFILE_SOURCE"
---
>   INTERFACE_COMPILE_DEFINITIONS "_POSIX_C_SOURCE=200809L;_GNU_SOURCE"
```

Also `CMakeFiles/Makefile.cmake` files are different, diff output is big, but e.g. with zlib mangling:
```
diff CMakeFiles-1/Makefile.cmake CMakeFiles-2/Makefile.cmake
...
<   "/home/r/itk/ITK/Modules/ThirdParty/ZLIB/src/itkzlib-ng/zlib_name_mangling.h.empty"
---
>   "/home/r/itk/ITK/Modules/ThirdParty/ZLIB/src/itkzlib-ng/zlib_name_mangling.h.in"
...
```

Probably I shall look again later and try to find out where `-msse2` is from, is it ITK or 3rd party.

File: CMake/ITKSetStandardCompilerFlags.cmake [hunk 2]
```diff
@@ -348,10 +348,6 @@ macro(check_compiler_platform_flags)
       set(ITK_REQUIRED_CXX_FLAGS "${ITK_REQUIRED_CXX_FLAGS} -fno-sized-deallocation")
     endif()
 
-    # gcc must have -msse2 option to enable sse2 support
```
[the remark this answers is not in the document]
Comment (maxaehle) [reply CMake/ITKSetStandardCompilerFlags.cmake:351]: As @issakomi also [answered](https://github.com/InsightSoftwareConsortium/ITK/pull/3852#issuecomment-1373320579), there is no need to provide the `-msse2` flag to GCC on x86-64. The deleted lines are below an `if(CMAKE_COMPILER_IS_GNUCXX)` check and the [GCC online documentation](https://gcc.gnu.org/onlinedocs/gcc/x86-Options.html) says that

> For the x86-32 compiler, you must use -march=cpu-type, -msse or -msse2 switches to enable SSE extensions and make this option effective. For the x86-64 compiler, these extensions are enabled by default. 

If you would like to provide the `-msse2` flag to GCC on 32-bit x86, I could have a look into how to do this, but I think it's not really worth the effort.

**Comment by maxaehle** [comment]
> Probably I shall look again later and try to find out where `-msse2` is from, is it ITK or 3rd party.

As far as I reconstructed it, `-msse2` flag is added in the ITK CMake code [here](https://github.com/InsightSoftwareConsortium/ITK/blob/06a6761d9e93f14e219ebc0243c31cd57f77b14a/CMake/ITKSetStandardCompilerFlags.cmake#L353) to `ITK_REQUIRED_CXX_FLAGS`, which is appended to `CMAKE_CXX_FLAGS` [here](https://github.com/InsightSoftwareConsortium/ITK/blob/06a6761d9e93f14e219ebc0243c31cd57f77b14a/CMakeLists.txt#L238).

Whether the above code adds `-msse2`  is controlled by the third-party option `VNL_CONFIG_ENABLE_SSE2_ROUNDING` (among others). This option is declared in the third-party code [here](https://github.com/InsightSoftwareConsortium/ITK/blob/06a6761d9e93f14e219ebc0243c31cd57f77b14a/Modules/ThirdParty/VNL/src/vxl/core/vnl/CMakeLists.txt#L31) with a default value `${VXL_HAS_SSE2_HARDWARE_SUPPORT}` (defined [here](https://github.com/InsightSoftwareConsortium/ITK/blob/06a6761d9e93f14e219ebc0243c31cd57f77b14a/Modules/ThirdParty/VNL/src/vxl/config/cmake/config/VXLIntrospectionConfig.cmake#L429), that's probably a check whether the CPU supports SSE2).

The reason why `-msse2` is not added in the first CMake run probably is that the declaration of `VNL_CONFIG_ENABLE_SSE2_ROUNDING` is processed *after* it is used to check whether to add `-msse2`. As the value of the option is stored in a cache variable (see the [CMake doc](https://cmake.org/cmake/help/latest/command/option.html)), in the second CMake run it is already before the check whether to add `-msse2`.

**Comment by issakomi** [comment]
> ```
> diff CMakeFiles-1/Makefile.cmake CMakeFiles-2/Makefile.cmake
> ...
> <   "/home/r/itk/ITK/Modules/ThirdParty/ZLIB/src/itkzlib-ng/zlib_name_mangling.h.empty"
> ---
> >   "/home/r/itk/ITK/Modules/ThirdParty/ZLIB/src/itkzlib-ng/zlib_name_mangling.h.in"
> ...
> ```

It looks like that after the 1st run _zlib_ mangling didn't work at all. There is **no** prefix in _libitkzlib-5.4.a_. After the 2nd run there is the prefix (at least partially). That stuff should be reviewed.

**1st run**

```
r@deb2:~/Desktop/1/b4/lib$ nm libitkzlib-5.4.a |grep " [TRD] "
0000000000000270 T adler32
0000000000000000 T adler32_c
00000000000002a0 T adler32_combine
0000000000000360 T adler32_combine64
0000000000000240 T adler32_z
0000000000000000 T adler32_fold_copy_c
0000000000000010 T chunkcopy_c
0000000000000090 T chunkmemset_c
0000000000000220 T chunkmemset_safe_c
0000000000000000 T chunksize_c
0000000000000050 T chunkunroll_c
0000000000000000 T compare256_c
0000000000000a70 T compare256_unaligned_16
0000000000001430 T compare256_unaligned_32
0000000000001c10 T compare256_unaligned_64
00000000000000a0 T longest_match_c
0000000000000480 T longest_match_slow_c
0000000000000e80 T longest_match_slow_unaligned_16
0000000000001700 T longest_match_slow_unaligned_32
0000000000001ef0 T longest_match_slow_unaligned_64
0000000000000b10 T longest_match_unaligned_16
0000000000001470 T longest_match_unaligned_32
0000000000001c50 T longest_match_unaligned_64
0000000000000110 T compress
0000000000000000 T compress2
0000000000000120 T compressBound
0000000000000000 T cpu_check_features
0000000000000050 T crc32
0000000000000060 T crc32_braid
0000000000000010 T crc32_z
0000000000000000 T get_crc_table
0000000000000000 T crc32_combine
00000000000000c0 T crc32_combine64
0000000000000180 T crc32_combine_gen
0000000000000200 T crc32_combine_gen64
0000000000000280 T crc32_combine_op
0000000000000070 T crc32_fold_c
0000000000000010 T crc32_fold_copy_c
00000000000000c0 T crc32_fold_final_c
0000000000000000 T crc32_fold_reset_c
0000000000000740 T deflate
00000000000005a0 T deflateBound
0000000000001780 T deflateCopy
0000000000000000 R deflate_copyright
0000000000001350 T deflateEnd
00000000000001c0 T deflateGetDictionary
0000000000001710 T deflateInit_
0000000000001420 T deflateInit2
0000000000001750 T deflateInit2_
00000000000011d0 T deflateParams
0000000000000410 T deflatePending
0000000000000460 T deflatePrime
0000000000000340 T deflateReset
0000000000000250 T deflateResetKeep
0000000000001d40 T deflateSetDictionary
00000000000003d0 T deflateSetHeader
0000000000000550 T deflateTune
0000000000001a80 T fill_window
00000000000006c0 T flush_pending
00000000000019c0 T read_buf
0000000000000000 T deflate_fast
0000000000000000 T deflate_huff
0000000000000000 T deflate_medium
0000000000000000 T deflate_quick
0000000000000000 T deflate_rle
0000000000000000 T deflate_slow
0000000000000000 T deflate_stored
0000000000000140 T adler32_fold_copy_stub
00000000000005a0 T adler32_stub
00000000000001e0 T chunkcopy_stub
0000000000000340 T chunkmemset_safe_stub
00000000000002c0 T chunkmemset_stub
0000000000000850 T chunksize_stub
0000000000000250 T chunkunroll_stub
00000000000003c0 T compare256_stub
0000000000000700 T crc32_fold_copy_stub
00000000000007e0 T crc32_fold_final_stub
0000000000000690 T crc32_fold_reset_stub
0000000000000770 T crc32_fold_stub
00000000000008b0 T crc32_stub
0000000000000000 D functable
00000000000004b0 T insert_string_stub
00000000000000d0 T longest_match_slow_stub
0000000000000060 T longest_match_stub
0000000000000000 T quick_insert_string_stub
0000000000000520 T slide_hash_stub
0000000000000430 T update_hash_stub
0000000000000140 T inflateBack
0000000000001300 T inflateBackEnd
0000000000000000 T inflateBackInit
0000000000000120 T inflateBackInit_
00000000000001a0 T zng_inflate_fast
0000000000000420 T fixedtables
00000000000008c0 T inflate
00000000000032b0 T inflateCodesUsed
0000000000002ef0 T inflateCopy
00000000000029b0 T inflateEnd
0000000000000450 T inflate_ensure_window
0000000000002a10 T inflateGetDictionary
0000000000002b80 T inflateGetHeader
0000000000000330 T inflateInit_
0000000000000210 T inflateInit2
0000000000000360 T inflateInit2_
0000000000003220 T inflateMark
0000000000000380 T inflatePrime
0000000000000110 T inflateReset
0000000000000170 T inflateReset2
0000000000000050 T inflateResetKeep
0000000000002ac0 T inflateSetDictionary
0000000000002bf0 T inflateSync
0000000000002e90 T inflateSyncPoint
0000000000003150 T inflateUndermine
00000000000031b0 T inflateValidate
0000000000000100 R inflate_copyright
0000000000000000 T zng_inflate_table
0000000000000050 T insert_string_c
0000000000000010 T quick_insert_string_c
0000000000000000 T update_hash_c
0000000000000050 T insert_string_roll
0000000000000010 T quick_insert_string_roll
0000000000000000 T update_hash_roll
0000000000000000 T slide_hash_c
0000000000000000 R base_dist
0000000000000080 R base_length
0000000000001c70 T bi_reverse
0000000000000d60 T gen_codes
0000000000000400 R static_dtree
0000000000000480 R static_ltree
0000000000000200 R zng_dist_code
0000000000000100 R zng_length_code
0000000000001450 T zng_tr_align
0000000000001440 T zng_tr_flush_bits
0000000000001550 T zng_tr_flush_block
0000000000000c70 T zng_tr_init
0000000000001350 T zng_tr_stored_block
00000000000001a0 T uncompress
0000000000000000 T uncompress2
0000000000000000 D z_errmsg
0000000000000070 T zError
0000000000000060 T zlibCompileFlags
0000000000000000 R zlibng_string
0000000000000050 T zlibng_version
0000000000000040 T zlibVersion
0000000000000090 T zng_alloc_aligned
0000000000000010 T zng_calloc
0000000000000000 T zng_cfree
0000000000000120 T zng_free_aligned
0000000000000000 T x86_check_features
0000000000000000 T slide_hash_avx2
0000000000000010 T chunkcopy_avx
00000000000000a0 T chunkmemset_avx
00000000000003a0 T chunkmemset_safe_avx
0000000000000000 T chunksize_avx
0000000000000050 T chunkunroll_avx
0000000000000000 T compare256_avx2
0000000000000060 T longest_match_avx2
0000000000000370 T longest_match_slow_avx2
0000000000000000 T adler32_avx2
00000000000001c0 T adler32_fold_copy_avx2
0000000000000000 T adler32_avx512
0000000000000190 T adler32_fold_copy_avx512
0000000000000000 T adler32_avx512_vnni
0000000000000220 T adler32_fold_copy_avx512_vnni
0000000000000210 T chunkmemset_safe_sse41
0000000000000010 T chunkmemset_sse41
0000000000000000 T chunksize_sse41
0000000000000000 T adler32_fold_copy_sse42
0000000000000050 T insert_string_sse4
0000000000000010 T quick_insert_string_sse4
0000000000000000 T update_hash_sse4
0000000000000010 T chunkcopy_sse2
0000000000000380 T chunkmemset_safe_sse2
0000000000000090 T chunkmemset_sse2
0000000000000000 T chunksize_sse2
0000000000000050 T chunkunroll_sse2
0000000000000000 T compare256_sse2
0000000000000410 T longest_match_slow_sse2
00000000000000b0 T longest_match_sse2
0000000000000000 T slide_hash_sse2
0000000000000000 T adler32_ssse3
00000000000000f0 T crc32_fold_pclmulqdq
0000000000000650 T crc32_fold_pclmulqdq_copy
0000000000000c00 T crc32_fold_pclmulqdq_final
00000000000000d0 T crc32_fold_pclmulqdq_reset
0000000000000cd0 T crc32_pclmulqdq
0000000000000000 T fold_16_vpclmulqdq
00000000000001b0 T fold_16_vpclmulqdq_copy
0000000000000420 T gzbuffer
0000000000000810 T gzclearerr
00000000000003f0 T gzclose
0000000000000380 T gzdopen
00000000000007a0 T gzeof
0000000000000890 T gz_error
00000000000007c0 T gzerror
0000000000000790 T gzoffset
0000000000000730 T gzoffset64
0000000000000360 T gzopen
0000000000000370 T gzopen64
0000000000000460 T gzrewind
00000000000006d0 T gzseek
0000000000000530 T gzseek64
0000000000000720 T gztell
00000000000006e0 T gztell64
0000000000000b90 T gzclose_r
0000000000000b40 T gzdirect
00000000000007a0 T gzfread
0000000000000820 T gzgetc
00000000000008a0 T gzgetc_
00000000000009e0 T gzgets
0000000000000740 T gzread
00000000000008b0 T gzungetc
0000000000000b40 T gzclose_w
00000000000009f0 T gzflush
0000000000000610 T gzfwrite
0000000000000960 T gzprintf
0000000000000680 T gzputc
0000000000000760 T gzputs
0000000000000a60 T gzsetparams
00000000000007e0 T gzvprintf
00000000000005c0 T gzwrite
```

-------------------------------------------------------------------------------------------------------------------------------------------

**2nd run:**



```
r@deb2:~/Desktop/2/b4/lib$ nm libitkzlib-5.4.a |grep " [TRD] "
0000000000000000 T adler32_c
0000000000000270 T itkzlib_adler32
00000000000002a0 T itkzlib_adler32_combine
0000000000000360 T itkzlib_adler32_combine64
0000000000000240 T itkzlib_adler32_z
0000000000000000 T adler32_fold_copy_c
0000000000000010 T chunkcopy_c
0000000000000090 T chunkmemset_c
0000000000000220 T chunkmemset_safe_c
0000000000000000 T chunksize_c
0000000000000050 T chunkunroll_c
0000000000000000 T compare256_c
0000000000000a70 T compare256_unaligned_16
0000000000001430 T compare256_unaligned_32
0000000000001c10 T compare256_unaligned_64
00000000000000a0 T longest_match_c
0000000000000480 T longest_match_slow_c
0000000000000e80 T longest_match_slow_unaligned_16
0000000000001700 T longest_match_slow_unaligned_32
0000000000001ef0 T longest_match_slow_unaligned_64
0000000000000b10 T longest_match_unaligned_16
0000000000001470 T longest_match_unaligned_32
0000000000001c50 T longest_match_unaligned_64
0000000000000110 T itkzlib_compress
0000000000000000 T itkzlib_compress2
0000000000000120 T itkzlib_compressBound
0000000000000000 T cpu_check_features
0000000000000060 T crc32_braid
0000000000000050 T itkzlib_crc32
0000000000000010 T itkzlib_crc32_z
0000000000000000 T itkzlib_get_crc_table
0000000000000000 T itkzlib_crc32_combine
00000000000000c0 T itkzlib_crc32_combine64
0000000000000180 T itkzlib_crc32_combine_gen
0000000000000200 T itkzlib_crc32_combine_gen64
0000000000000280 T itkzlib_crc32_combine_op
0000000000000070 T crc32_fold_c
0000000000000010 T crc32_fold_copy_c
00000000000000c0 T crc32_fold_final_c
0000000000000000 T crc32_fold_reset_c
0000000000001420 T deflateInit2
0000000000001a80 T fill_window
00000000000006c0 T flush_pending
0000000000000740 T itkzlib_deflate
00000000000005a0 T itkzlib_deflateBound
0000000000001780 T itkzlib_deflateCopy
0000000000000000 R itkzlib_deflate_copyright
0000000000001350 T itkzlib_deflateEnd
00000000000001c0 T itkzlib_deflateGetDictionary
0000000000001710 T itkzlib_deflateInit_
0000000000001750 T itkzlib_deflateInit2_
00000000000011d0 T itkzlib_deflateParams
0000000000000410 T itkzlib_deflatePending
0000000000000460 T itkzlib_deflatePrime
0000000000000340 T itkzlib_deflateReset
0000000000000250 T itkzlib_deflateResetKeep
0000000000001d40 T itkzlib_deflateSetDictionary
00000000000003d0 T itkzlib_deflateSetHeader
0000000000000550 T itkzlib_deflateTune
00000000000019c0 T read_buf
0000000000000000 T deflate_fast
0000000000000000 T deflate_huff
0000000000000000 T deflate_medium
0000000000000000 T deflate_quick
0000000000000000 T deflate_rle
0000000000000000 T deflate_slow
0000000000000000 T deflate_stored
0000000000000140 T adler32_fold_copy_stub
00000000000005a0 T adler32_stub
00000000000001e0 T chunkcopy_stub
0000000000000340 T chunkmemset_safe_stub
00000000000002c0 T chunkmemset_stub
0000000000000850 T chunksize_stub
0000000000000250 T chunkunroll_stub
00000000000003c0 T compare256_stub
0000000000000700 T crc32_fold_copy_stub
00000000000007e0 T crc32_fold_final_stub
0000000000000690 T crc32_fold_reset_stub
0000000000000770 T crc32_fold_stub
00000000000008b0 T crc32_stub
0000000000000000 D functable
00000000000004b0 T insert_string_stub
00000000000000d0 T longest_match_slow_stub
0000000000000060 T longest_match_stub
0000000000000000 T quick_insert_string_stub
0000000000000520 T slide_hash_stub
0000000000000430 T update_hash_stub
0000000000000000 T inflateBackInit
0000000000000140 T itkzlib_inflateBack
0000000000001300 T itkzlib_inflateBackEnd
0000000000000120 T itkzlib_inflateBackInit_
00000000000001a0 T zng_inflate_fast
0000000000000420 T fixedtables
0000000000000450 T inflate_ensure_window
0000000000000210 T inflateInit2
00000000000008c0 T itkzlib_inflate
00000000000032b0 T itkzlib_inflateCodesUsed
0000000000002ef0 T itkzlib_inflateCopy
00000000000029b0 T itkzlib_inflateEnd
0000000000002a10 T itkzlib_inflateGetDictionary
0000000000002b80 T itkzlib_inflateGetHeader
0000000000000330 T itkzlib_inflateInit_
0000000000000360 T itkzlib_inflateInit2_
0000000000003220 T itkzlib_inflateMark
0000000000000380 T itkzlib_inflatePrime
0000000000000110 T itkzlib_inflateReset
0000000000000170 T itkzlib_inflateReset2
0000000000000050 T itkzlib_inflateResetKeep
0000000000002ac0 T itkzlib_inflateSetDictionary
0000000000002bf0 T itkzlib_inflateSync
0000000000002e90 T itkzlib_inflateSyncPoint
0000000000003150 T itkzlib_inflateUndermine
00000000000031b0 T itkzlib_inflateValidate
0000000000000100 R itkzlib_inflate_copyright
0000000000000000 T zng_inflate_table
0000000000000050 T insert_string_c
0000000000000010 T quick_insert_string_c
0000000000000000 T update_hash_c
0000000000000050 T insert_string_roll
0000000000000010 T quick_insert_string_roll
0000000000000000 T update_hash_roll
0000000000000000 T slide_hash_c
0000000000000000 R base_dist
0000000000000080 R base_length
0000000000001c70 T bi_reverse
0000000000000d60 T gen_codes
0000000000000400 R static_dtree
0000000000000480 R static_ltree
0000000000000200 R zng_dist_code
0000000000000100 R zng_length_code
0000000000001450 T zng_tr_align
0000000000001440 T zng_tr_flush_bits
0000000000001550 T zng_tr_flush_block
0000000000000c70 T zng_tr_init
0000000000001350 T zng_tr_stored_block
00000000000001a0 T itkzlib_uncompress
0000000000000000 T itkzlib_uncompress2
0000000000000000 D itkzlib_z_errmsg
0000000000000070 T itkzlib_zError
0000000000000060 T itkzlib_zlibCompileFlags
0000000000000040 T itkzlib_zlibVersion
0000000000000000 R zlibng_string
0000000000000050 T zlibng_version
0000000000000090 T zng_alloc_aligned
0000000000000010 T zng_calloc
0000000000000000 T zng_cfree
0000000000000120 T zng_free_aligned
0000000000000000 T x86_check_features
0000000000000000 T slide_hash_avx2
0000000000000010 T chunkcopy_avx
00000000000000a0 T chunkmemset_avx
00000000000003a0 T chunkmemset_safe_avx
0000000000000000 T chunksize_avx
0000000000000050 T chunkunroll_avx
0000000000000000 T compare256_avx2
0000000000000060 T longest_match_avx2
0000000000000370 T longest_match_slow_avx2
0000000000000000 T adler32_avx2
00000000000001c0 T adler32_fold_copy_avx2
0000000000000000 T adler32_avx512
0000000000000190 T adler32_fold_copy_avx512
0000000000000000 T adler32_avx512_vnni
0000000000000220 T adler32_fold_copy_avx512_vnni
0000000000000210 T chunkmemset_safe_sse41
0000000000000010 T chunkmemset_sse41
0000000000000000 T chunksize_sse41
0000000000000000 T adler32_fold_copy_sse42
0000000000000050 T insert_string_sse4
0000000000000010 T quick_insert_string_sse4
0000000000000000 T update_hash_sse4
0000000000000010 T chunkcopy_sse2
0000000000000380 T chunkmemset_safe_sse2
0000000000000090 T chunkmemset_sse2
0000000000000000 T chunksize_sse2
0000000000000050 T chunkunroll_sse2
0000000000000000 T compare256_sse2
0000000000000410 T longest_match_slow_sse2
00000000000000b0 T longest_match_sse2
0000000000000000 T slide_hash_sse2
0000000000000000 T adler32_ssse3
00000000000000f0 T crc32_fold_pclmulqdq
0000000000000650 T crc32_fold_pclmulqdq_copy
0000000000000c00 T crc32_fold_pclmulqdq_final
00000000000000d0 T crc32_fold_pclmulqdq_reset
0000000000000cd0 T crc32_pclmulqdq
0000000000000000 T fold_16_vpclmulqdq
00000000000001b0 T fold_16_vpclmulqdq_copy
0000000000000420 T itkzlib_gzbuffer
0000000000000810 T itkzlib_gzclearerr
00000000000003f0 T itkzlib_gzclose
0000000000000380 T itkzlib_gzdopen
00000000000007a0 T itkzlib_gzeof
0000000000000890 T itkzlib_gz_error
00000000000007c0 T itkzlib_gzerror
0000000000000790 T itkzlib_gzoffset
0000000000000730 T itkzlib_gzoffset64
0000000000000360 T itkzlib_gzopen
0000000000000370 T itkzlib_gzopen64
0000000000000460 T itkzlib_gzrewind
00000000000006d0 T itkzlib_gzseek
0000000000000530 T itkzlib_gzseek64
0000000000000720 T itkzlib_gztell
00000000000006e0 T itkzlib_gztell64
0000000000000b90 T itkzlib_gzclose_r
0000000000000b40 T itkzlib_gzdirect
00000000000007a0 T itkzlib_gzfread
0000000000000820 T itkzlib_gzgetc
00000000000008a0 T itkzlib_gzgetc_
00000000000009e0 T itkzlib_gzgets
0000000000000740 T itkzlib_gzread
00000000000008b0 T itkzlib_gzungetc
0000000000000b40 T itkzlib_gzclose_w
00000000000009f0 T itkzlib_gzflush
0000000000000610 T itkzlib_gzfwrite
0000000000000960 T itkzlib_gzprintf
0000000000000680 T itkzlib_gzputc
0000000000000760 T itkzlib_gzputs
0000000000000a60 T itkzlib_gzsetparams
00000000000007e0 T itkzlib_gzvprintf
00000000000005c0 T itkzlib_gzwrite
```

File: Modules/ThirdParty/ZLIB/src/CMakeLists.txt [hunk 3]
```diff
@@ -2,7 +2,7 @@ set(INC_INSTALL_DIR "${ITKZLIB_INSTALL_INCLUDE_DIR}/itkzlib-ng")
 set(BIN_INSTALL_DIR "${ITKZLIB_INSTALL_RUNTIME_DIR}")
 set(LIB_INSTALL_DIR "${ITKZLIB_INSTALL_LIBRARY_DIR}")
 set(PKGCONFIG_INSTALL_DIR "${ITKZLIB_INSTALL_ARCHIVE_DIR}")
-set(ZLIB_SYMBOL_PREFIX "itkzlib_")
+set(ZLIB_SYMBOL_PREFIX "itkzlib_" CACHE INTERNAL "Hardcoded prefix for symbols in zlib")
```
[the remark this answers is not in the document]
Comment (maxaehle) [reply Modules/ThirdParty/ZLIB/src/CMakeLists.txt:5]: @issakomi verified it [below](https://github.com/InsightSoftwareConsortium/ITK/pull/3852#issuecomment-1373408278).

[the remark this answers is not in the document]
Comment (issakomi) [reply Modules/ThirdParty/ZLIB/src/CMakeLists.txt:5]: The above change fixed the issue.

File: Modules/ThirdParty/HDF5/src/itkhdf5/config/cmake_ext_mod/ConfigureChecks.cmake (hunk 1 above)
Comment (maxaehle) [reply Modules/ThirdParty/HDF5/src/itkhdf5/config/cmake_ext_mod/ConfigureChecks.cmake:278]: I created a new issue https://github.com/HDFGroup/hdf5/issues/2395 in HDF5.

File: CMake/ITKSetStandardCompilerFlags.cmake (hunk 2 above)
[the remark this answers is not in the document]
Comment (thewtex) [reply CMake/ITKSetStandardCompilerFlags.cmake:351]: @maxaehle thanks for working on this contribution!

> For the x86-32 compiler, you must use -march=cpu-type, -msse or -msse2 switches to enable SSE extensions and make this option effective. For the x86-64 compiler, these extensions are enabled by default.

From what I can see, it seems that this only applies when

> -mfpmath=see

is set. Is this correct?

[the remark this answers is not in the document]
Comment (issakomi) [reply CMake/ITKSetStandardCompilerFlags.cmake:351]: The `-mfpmath` could be related to specialty of Pentium III's SSE implementation, s. [wikipedia](https://en.wikipedia.org/wiki/Pentium_III), section _Pentium III's SSE implementation_. Not 100% sure. AFAIK, on new x86-32 CPU `-msse2` is enough.
It is a little bit difficult part, neither ITK's dashboard nor my personal needs require x86-32 SSE2 support, but there are still users, AFAIK. There have been recent support requests related to the x86-32 build.

[the remark this answers is not in the document]
Comment (issakomi) [reply CMake/ITKSetStandardCompilerFlags.cmake:351]: Maybe a kind of workaround? It is not elegant, but probably will work on x86-64. There still may be an issue on x86-32, not sure.

```
+ if (CMAKE_SIZEOF_VOID_P EQUAL 4)
     if(VNL_CONFIG_ENABLE_SSE2 OR VNL_CONFIG_ENABLE_SSE2_ROUNDING)
       set(ITK_REQUIRED_CXX_FLAGS "${ITK_REQUIRED_CXX_FLAGS} -msse2")
     endif()
+ endif()
```

Please wait for @dzenanz response before spending your time with the change.

[the remark this answers is not in the document]
Comment (issakomi) [reply CMake/ITKSetStandardCompilerFlags.cmake:351]: > Copy the test ... and thus evaluated "too late", to ITK CMake code

For the record, there was the try-compile test in the past, it was removed with the commit 
https://github.com/InsightSoftwareConsortium/ITK/commit/0402e501508c3b1134b222de43e269935091b88c
```
COMP: remove try-compile for SSE2 detection
Instead rely on various preprocessor macros to detect if SSE2 is available.
```
BTW, i think that "preprocessor macros" (`__SSE2__` definition) depend on `-msse2` argument on x86-32.

If there will be more changes, i could try to test in VMware (I have recently deleted x86-32 Linux VM, but can install again little later for tests).

**Comment by dzenanz** [comment]
@maxaehle If you do not plan to tackle SSE flag soon, maybe remove it from this PR? It can be always done later. Making an issue to track the problem is good for keeping it in mind.

`zlib` and `hdf5` changes seem ready.

**Comment by maxaehle** [comment]
I'm sorry for keeping you waiting. In principle I would propose to conditionally add the `-msse2` flag using this patch: 
[ITK-idempotent-cmake-continuation-sse2-diff.txt](https://github.com/InsightSoftwareConsortium/ITK/files/10509806/ITK-idempotent-cmake-continuation-sse2-diff.txt). However, I am not very experienced in CMake and I have only tested this patch on an Ubuntu system with GCC 9.4.0 and CMake 3.16.3 on x86_64, so it would require much more testing by others.

I'm both fine with merging the zlib and hdf5 changes now and postponing the SSE2 change to another issue/PR, and with continuing in this PR. If we continue here, I'll commit the above patch to this branch.

**Comment by dzenanz** [comment]
If you have the patch already, please add it to this PR.

**Comment by dzenanz** [comment]
Could you squash the latest commit with the second one "COMP: Disable global GCC flag -msse"? The squashed commit could be last, after zlib and hdf5 commits.

**Review by thewtex — APPROVED** [review]
@maxaehle thank you!

**Pull request merged**